File: Math&Geometry/3443. Maximum Manhattan Distance After K Changes/solution.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        N = S = E = W = res = 0
        for d in s:
            if d == "N":
                N += 1
            elif d == "S":
                S += 1
            elif d == "E":
                E += 1
            else:
                W += 1

            x1, y1 = max(N, S), min(N, S)
            x2, y2 = max(E, W), min(E, W)

            kk = k
            d1 = min(y1, kk)
            y1 -= d1
            kk -= d1

            d2 = min(y2, kk)
            y2 -= d2
            kk -= d2

            res = max(res, x1 + x2 + d1 + d2 - y1 - y2)
        return res

# greedily replace lower frequency direction and calculate maximum Manhattan distance.
# each replacement can gain 1 more distince

File: Math&Geometry/3443. Maximum Manhattan Distance After K Changes/solution.cpp
class Solution {
public:
    int maxDistance(string s, int k) {
        int x=0, y=0;
        int N=0, S=0, E=0, W=0;
        int res = 0;
        for (int i=0; i < s.size(); i++) {
            if (s[i] == 'N') {
                y += 1;
                N += 1;
            } else if (s[i] == 'S') {
                y -= 1;
                S += 1;
            } else if (s[i] == 'E') {
                x += 1;
                E += 1;
            } else {
                x -= 1;
                W += 1;
            }

            int dist = abs(x) + abs(y);
            res = max(res, dist);
            
            int xx = min(E, W);
            int yy = min(N, S);
            res = max(res, dist + 2 * min(xx+yy, k)); // use k operations to replace minimum opposite direction and we can gain `2k` at most
        }
        return res;
    }
};